app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsFragment.kt

```diff
@@ -199,11 +200,11 @@ class AccountSettingsFragment : PreferenceFragmentCompat(), ConfirmationDialogFr
 
     private fun initializeCryptoSettings(account: Account) {
         findPreference(PREFERENCE_OPENPGP)?.let {
-            configureCryptoPreferences(account)
+            configureCryptoPreferences(account, false)
         }
     }
 
-    private fun configureCryptoPreferences(account: Account) {
+    private fun configureCryptoPreferences(account: Account, needSave: Boolean) {
```

Comment:
I think having a boolean parameter like this makes it much harder to follow what's going on.

Maybe it helps to take a step back. The fundamental issue that needs to be fixed is the fact that some parts of this class modify state in `Account` without saving the account, i.e. write the changes to disk. While this is an indicator that that code should probably live somewhere else (all the other code that modifies `Account` lives in `AccountSettingsDataStore`), let's focus on fixing just the saving part.
My preferred way of doing this, is to be as explicit as possible. So every time `account.openPgpProvider` is written to, extract that code to a method that additionally also calls `preferences.saveAccount(account)`. I ended up with these methods:

```kotlin
private fun setOpenPgpProvider(account: Account, openPgpProviderPackage: String) {
    account.openPgpProvider = openPgpProviderPackage
    saveAccountInBackground(account)
}

private fun removeOpenPgpProvider(account: Account) {
    account.openPgpProvider = null
    account.openPgpKey = Account.NO_OPENPGP_KEY
    saveAccountInBackground(account)
}
```

As the method name `saveAccountInBackground()` suggests, we also want to perform the actual save operation in a background thread, not in the main thread.

Reply:
Good point! I found another bug with OpenPGP pane today: if OpenPGP is turned on from off-state and the key is selected, it gets written to database but then account settings do overwrite the good `keyId` with 0. I did have time today to investigate it deeply, but it seems like the another brick in tbe same wall.